jiant/tasks/tasks.py

```diff
@@ -3731,3 +3732,160 @@ def get_metrics(self, reset=False):
+    """
+
+    def __init__(self, path, max_seq_len, name, **kw):
+        super(SentenceOrderTask, self).__init__(name, n_classes=2, **kw)
+        self.path = path
+        self.max_seq_len = max_seq_len
+        self.train_data_text = None
+        self.val_data_text = None
+        self.test_data_text = None
+        self.files_by_split = {
+            "train": os.path.join(path, "train.txt"),
+            "val": os.path.join(path, "valid.txt"),
+            "test": os.path.join(path, "test.txt"),
+        }
+        self._label_namespace = self.name + "_labels"
+
+    def get_target_seq_length(self):
+        target_is_max = random.random() > 0.1
+        if target_is_max:
+            target_seq_length = self.max_seq_len
+        else:
+            target_seq_length = random.randint(2, self.max_seq_len)
+        return target_seq_length
+
+    def get_data_iter(self, path):
+        """Loading data file and tokenizing the text. We override the
+        this function and all functions that call this function because
+        the step of reading in the data for SOP is different than other
+        PairClassificationTasks.
+
+        ALBERT does SOP classification by, for each document:
+            For each example:
+                -90% of the time, getting sentences of max_seq_length number of tokens, and
+                10% of the time, getting sentences of a random number of tokens between 2 and max_seq_length.
+                -Given the sampled sentences, randomly sample N such that the first N sentences in the 
+                sampled go to the first segment, and the rest go to the second. 
+                -50% of the time, the first and second segments are switched. 
+            There is also some logic here to make sure that we always split by the sentence instead of 
+            in the middle of a sentence.
+        Args:
+            path: (str) data file path
+        """
+        f = open(path, "r")
+        #  The dataset comes with one sentence per line, thus we split by
+        #  line here.
+        current_chunk = [tokenize_and_truncate(self._tokenizer_name, next(f), -1)]
+        current_length = 0
+        target_seq_length = self.get_target_seq_length()
+        in_order = 1
+        while len(current_chunk) > 0:
+            segment = next(f)
+            segment = tokenize_and_truncate(self._tokenizer_name, segment, -1)
+            if "END OF ARTICLE" in segment or current_length >= target_seq_length:
+                for_next_chunk = []
+                if "END OF ARTICLE" not in segment:
+                    for_next_chunk.append(segment)
+                if current_length > target_seq_length:
+                    # Since the most current sentence added to the chunk exceeds the target
+                    # length, we save it for the next chunk (next example).
+                    for_next_chunk.append(current_chunk[-1])
```

Comment:
Should be done in the reverse order. Put the part that's too long from the end of `current_chunks` into `for_next_chunk`, then add the current `segment` to `for_next_chunk`.